Source/CarthageKit/Project.swift

```diff
@@ -581,25 +581,46 @@ public final class Project {
 		return cloneOrFetchDependency(project, commitish: revision)
 			.flatMap(.merge) { repositoryURL -> SignalProducer<(), CarthageError> in
 				let workingDirectoryURL = self.directoryURL.appendingPathComponent(project.relativePath, isDirectory: true)
-				var submodule: Submodule?
 				
-				if var foundSubmodule = submodulesByPath[project.relativePath] {
-					foundSubmodule.url = repositoryURLForProject(project, preferHTTPS: self.preferHTTPS)
-					foundSubmodule.sha = revision
-					submodule = foundSubmodule
-				} else if self.useSubmodules {
-					submodule = Submodule(name: project.relativePath, path: project.relativePath, url: repositoryURLForProject(project, preferHTTPS: self.preferHTTPS), sha: revision)
+				/// The submodule for an already existing submodule at dependency project’s path
+				/// or the submodule to be added at this path given the `--use-submodules` flag.
+				func submodule() -> Submodule? {
+					if var foundSubmodule = submodulesByPath[project.relativePath] {
+						foundSubmodule.url = repositoryURLForProject(project, preferHTTPS: self.preferHTTPS)
+						foundSubmodule.sha = revision
+						return foundSubmodule
+					} else if self.useSubmodules {
+						return Submodule(name: project.relativePath, path: project.relativePath, url: repositoryURLForProject(project, preferHTTPS: self.preferHTTPS), sha: revision)
+					} else {
+						return nil
+					}
 				}
-				
-				if let submodule = submodule {
-					return addSubmoduleToRepository(self.directoryURL, submodule, GitURL(repositoryURL.carthage_path))
-						.startOnQueue(self.gitOperationQueue)
+
+				func symlinkCheckoutPaths(submodule submodule: Submodule?) -> SignalProducer<(), CarthageError> {
+					return self.symlinkCheckoutPathsForDependencyProject(
+						dependency,
+						repositoryURL: repositoryURL,
+						rootDirectoryURL: self.directoryURL,
+						// In the presense of `submodule` for `dependency` — before symlinking, (not after) — add submodule and its submodules:
+						// `dependency`, subdependencies that are submodules, and non-Carthage-housed submodules.
+						addSubmoduleForDependency: submodule.map {
+							addSubmoduleToRepository(self.directoryURL, $0, GitURL(repositoryURL.path!)).startOnQueue(self.gitOperationQueue)
```

Comment:
Would it make sense to move this logic into `symlinkCheckoutPathsForDependencyProject`?